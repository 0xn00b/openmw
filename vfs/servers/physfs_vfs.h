#ifndef MANGLE_VFS_PHYSFS_SERVER_H
#define MANGLE_VFS_PHYSFS_SERVER_H

#include "../vfs.h"
#include "../../stream/servers/phys_stream.h"

#include <physfs.h>
#include <assert.h>

namespace Mangle {
namespace VFS {

/** @brief An interface into the PhysFS virtual file system library

    You have to set up PhysFS on your own before using this class.
 */
class PhysVFS : public VFS
{
 public:
  PhysVFS()
    {
      hasList = true;
      hasFind = false;
      isCaseSensitive = true;
    }

  /// Open a new data stream. Deleting the object should be enough to
  /// close it.
  virtual Stream::Stream *open(const std::string &name)
    { return new Stream::PhysFile(PHYSFS_openRead(name.c_str())); }

  /// Check for the existence of a file
  virtual bool isFile(const std::string &name) const
    { return PHYSFS_exists(name.c_str()); }

  /// Checks for a directory
  virtual bool isDir(const std::string &name) const
    { return PHYSFS_isDirectory(name.c_str()); }

  /// This doesn't work
  virtual FileInfo stat(const std::string &name) const
    { assert(0); return FileInfo(); }

  virtual FileInfoList list(const std::string& dir = "",
                            bool recurse=true,
                            bool dirs=false) const
    {
      char **files = PHYSFS_enumerateFiles(dir.c_str());
      FileInfoList lst;

      // Add all teh files
      int i = 0;
      while(files[i] != NULL)
        {
          FileInfo fi;
          fi.name = files[i];
          fi.isDir = false;

          lst.push_back(fi);
        }
      return lst;
    }

  virtual FileInfoList find(const std::string& pattern,
                            bool recursive=true,
                            bool dirs=false) const
    { assert(0); }
};

}} // namespaces
#endif
