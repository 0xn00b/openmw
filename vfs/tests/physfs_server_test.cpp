#include "../servers/physfs_vfs.h"

#include "server_common.cpp"

#include <physfs.h>

int main()
{
  // Set up the library and paths
  PHYSFS_init("blah");
  PHYSFS_addToSearchPath("test.zip", 1);
  PHYSFS_addToSearchPath("./", 1);

  // Create our interface
  PhysVFS vfs;

  // Run the test
  testAll(vfs);

  return 0;
}
