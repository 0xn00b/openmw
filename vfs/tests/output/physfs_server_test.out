
File: Makefile
Size: 814
First 12 bytes: GCC=g++ -I..

File: testfile.txt
Size: 13
First 12 bytes: hello world!

File: blah_bleh
File doesn't exist
